2024/BIN-2024-0006.md

```diff
@@ -0,0 +1,156 @@
+
+### Implementation
+
+```c++
+case OP_PAIRCOMMIT: {
+    // OP_PAIRCOMMIT is only available in Tapscript
+    // ...
+    // x1 x2 -- hash
+    if (stack.size() < 2) {
+        return set_error(serror, SCRIPT_ERR_INVALID_STACK_OPERATION);
+    }
+    const valtype& vch1 = stacktop(-2);
+    const valtype& vch2 = stacktop(-1);
+
+    uint256 hash = PairCommitHash(vch1, vch2);
+
+    popstack(stack);
+    popstack(stack);
+    stack.push_back(ToByteVector(hash));
+    break;
+}
+```
+Serialization happens with little endian byte order as per default
+```c++
+const HashWriter HASHER_PAIRCOMMIT{TaggedHash("PairCommit")};
+
+uint256 PairCommitHash(Span<const unsigned char> x1, Span<const unsigned char> x2)
+{
+    // PAD is 0x00, 0x00, 0x00, 0x01 in little endian serializaton
+    static const uint32_t PCPAD = 0x01000000u;
+
+    HashWriter ss{HASHER_PAIRCOMMIT};
+    ss << x1
+       << x2
+       << uint32_t(x1.size()) << PCPAD
+       << uint32_t(x2.size()) << PCPAD;
+
+    return ss.GetSHA256();
+}
+```
+See more: https://github.com/lnhance/bitcoin/pull/6/files
+
+## Motivation
+
+Pair or vector committments are generally useful in a covenant script toolkit.
+Finding the simplest and safest way to do such data committments, and making sure
+to the outmost degree possible, that it would not intoduce unintended behavior,
+such as novel 2-way peg mechanisms was the primary motivation for this proposal.
+
+The number of SHA256 iterations is minimized in the most likely use case we can
+optimize for, which is LN-Symmetry. Since the Tag can be pre-computed as mid-state,
+it would only take 1 or 2 hash cycles in validation for the unilateral close
+scenario for Symmetry channels.
+
+In case of a 7 byte balance commitment + 32 byte CTV hash (no HTLCs in-flight),
+the total preimage size is 55 bytes. Which should make it fit into a single block
+with the SHA256 length commitment.
+
+In case there are 2x 32 byte CTV hash commitments, the first 64 byte block is
+comprised of those hashes, and the second block is the vectors’ size and total
+length commitment, which would be a largely 0 filled block with a very few bits
+set to 1.
+
+It’s a particular concern for LN-Symmetry with CTV that the concatenation of the
+two preimages allows for length redistribution attacks, because CTV is only defined
+for 32 byte templates and will act as NOP for different template sizes for
+upgradeability.
+
+### Vector Commitments
+
+`OP_PAIRCOMMIT` can be used to commit to a vector of stack elements in a way that
+is not vulnerable to various forms of witness malleability especially when used
+in conjunction with `OP_CHECKSIGFROMSTACK`[^CSFS] and `OP_INTERNALKEY`[^IKEY],
+making the script cleaner, and simpler. If `OP_CAT`[^CAT] was used naively, the
+contract could be easily broken since `OP_CHECKTEMPLATEVERIFY`[^CTV] is only
+defined for 32 byte parameters.
```

Comment:
Thinking more about on vector commitments, it actually is impossible to predict what would be more optimal for the user a) leave the items on the stack and only consume the number of items, or b) consume the items from the stack.

So i would probably do <vch1> .. <vchn> <n> VECTORCOMMIT where n as a signed char can be 2..127 or -2..-127, and if it's positive the stack elements are left on stack, if negative they are consumed. This would ofc be a different opcode and is not optimal for Symmetry.

SHA256 streaming could be a useful addition if we already have CAT with script restoration. I'm not sure if consensus is possible on that any time soon. In current script it would be too controversial as it naively relaxes the script limitations on what is possible both the limitation of stack element size that can get hashed with CAT only and without CAT it allows for custom construction of 'sighashes' like CTV templates or with CSFS pretty much everything CAT enables in terms of introspection.

PAIRCOMMIT is domain separated and limited on purpose, to hopefully avoid that controversy.